src/coreclr/vm/object.cpp

```diff
@@ -1906,40 +1899,119 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+// If the stack trace was created by another thread, it returns clones of both arrays.
+/* static */
+bool ExceptionObject::GetStackTraceParts(OBJECTREF stackTraceObj, StackTraceArray & stackTrace, PTRARRAYREF * outKeepaliveArray /*= NULL*/)
+{
+    CONTRACTL
+    {
+        GC_TRIGGERS;
+        THROWS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
 
-    if (outDynamicMethodArray != NULL)
+    struct
     {
-        *outDynamicMethodArray = _dynamicMethods;
+        StackTraceArray newStackTrace;
+        PTRARRAYREF keepaliveArray = NULL;
+        PTRARRAYREF newKeepaliveArray = NULL;
+        PTR_PTRArray combinedArray = NULL;
+    } gc;
+
+    bool wasCreatedByForeignThread = false;
+    
+    GCPROTECT_BEGIN(gc);
+
+    // Extract the stack trace and keepalive arrays from the stack trace object.
+    if ((stackTraceObj != NULL) && ((dac_cast<PTR_ArrayBase>(OBJECTREFToObject(stackTraceObj)))->GetArrayElementType() != ELEMENT_TYPE_I1))
+    {
+        // The stack trace object is the dynamic methods array with its first slot set to the stack trace I1Array.
+        gc.combinedArray = dac_cast<PTR_PTRArray>(OBJECTREFToObject(stackTraceObj));
+        stackTrace.Set(dac_cast<I1ARRAYREF>(gc.combinedArray->GetAt(0)));
+        gc.keepaliveArray = dac_cast<PTRARRAYREF>(ObjectToOBJECTREF(gc.combinedArray));
+    }
+    else
+    {
+        stackTrace.Set(dac_cast<I1ARRAYREF>(stackTraceObj));
     }
 
-#if !defined(DACCESS_COMPILE)
-    SpinLock::ReleaseLock(&g_StackTraceArrayLock);
-#endif // !defined(DACCESS_COMPILE)
+#ifndef DACCESS_COMPILE
+    Thread *pThread = GetThread();
+
+    if ((stackTrace.Get() != NULL) && (stackTrace.GetObjectThread() != pThread))
+    {
+        // When the stack trace was created by other thread than the current one, we create a copy of both the stack trace and the keepalive arrays to make sure
+        // they are not changing while the caller is accessing them.
+        wasCreatedByForeignThread = true;
+        gc.newStackTrace.Allocate(stackTrace.Capacity());
+        gc.newStackTrace.CopyDataFrom(stackTrace);
+        stackTrace.Set(gc.newStackTrace.Get());
+
+        int keepaliveArrayCapacity = (gc.keepaliveArray == NULL) ? 0 : gc.keepaliveArray->GetNumComponents();
+
+        int j = 0;
+        size_t count = stackTrace.Size();
+        for (size_t i = 0; i < count; i++)
+        {
+            MethodDesc *pMethod = stackTrace[i].pFunc;
+            if (pMethod->IsLCGMethod() || pMethod->GetMethodTable()->Collectible())
+            {
+                OBJECTREF keepaliveObject = NULL;
+                if ((j + 1) < keepaliveArrayCapacity)
+                {
+                    keepaliveObject = gc.keepaliveArray->GetAt(j + 1);
+                }
+                if (keepaliveObject == NULL)
+                {
+                    // Trim the stack trace at a point where a dynamic or collectible method is found without a corresponding keepalive object.
```

Comment:
How is this possible?

If I understand the code correctly, the thread safety should be guaranteed by writing the `size` field last and reading it first. Is that right? (Whatever it is, it would be nice to have it documented in a comment somewhere.)

I do not think it should be possible to get methods items without corresponding keep alive items here. If it was possible, I suspect there may be situations where the method is not kept alive and the `pMethod->IsLCGMethod()` check above can crash.

Reply:
Here is the case when it happens:
* Thread A owns the stack trace and is adding a new entry that requires a keep alive object.
* Thread B wants to update the stack trace too, so it fetches the stack trace array S1 and keep alive array K1 from the exception
* Thread A finds the keep alive array is full, so it creates a larger clone of the previous one, let's call it K2
* Thread A adds the new entry to keep alive array K2
* Thread A adds the new entry to stack trace array S1
* Thread A writes K2 to the exception
* Now thread B creates clones of S1 and K1 (when it read it from the exception, the K2 was not there yet). But S1 already has the new entry added by thread A, which is not covered by K1. Since K1 won't change anymore, we solve this by trimming the S1 at the element where we've found a frame that needs keep alive object and was not covered by K1

However, I have realized you are right that in such case, calling IsLCGMethod on the MethodTable above could crash. For example, if the thread A's exception was collected together with its keep alive array K2 that can be the only thing holding the method alive and if that would happen before the thread B called the IsLCGMethod.
So I need to figure out some other way to handle the situation described above.